### ISSUE TITLE
identity_op fails to deal with references properly: &0 + 0 

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#9904]
### Summary

.

### Reproducer

I tried this code:

we turn `&0 + 0` into `&0` but as per type annotation it should just be `0`. :/ 

```rust
fn main() {
    let _: u8 = &0 + 0;
}
```

`cargo clippy --fix -- -Aclippy::all -Wclippy::identity_op`

````
warning: failed to automatically apply fixes suggested by rustc to crate `clpy`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0308]: mismatched types
 --> src/main.rs:2:17
  |
2 |     let _: u8 = &0;
  |            --   ^^ expected `u8`, found `&{integer}`
  |            |
  |            expected due to this
  |
help: consider removing the borrow
  |
2 -     let _: u8 = &0;
2 +     let _: u8 = 0;
  |
help: alternatively, consider changing the type annotation
  |
2 |     let _: &u8 = &0;
  |            +

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
Original diagnostics will follow.

warning: this operation has no effect
 --> src/main.rs:2:17
  |
2 |     let _: u8 = &0 + 0;
  |                 ^^^^^^ help: consider reducing it to: `0`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#identity_op
  = note: requested on the command line with `-W clippy::identity-op`

warning: this operation has no effect
 --> src/main.rs:2:17
  |
2 |     let _: u8 = &0 + 0;
  |                 ^^^^^^ help: consider reducing it to: `&0`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#identity_op

warning: `clpy` (bin "clpy") generated 2 warnings (run `cargo fix --bin "clpy"` to apply 2 suggestions)
warning: `clpy` (bin "clpy" test) generated 2 warnings (2 duplicates)
    Finished dev [unoptimized + debuginfo] target(s) in 0.54s
````

### Version

```text
rustc 1.67.0-nightly (c5d82ed7a 2022-11-19)
binary: rustc
commit-hash: c5d82ed7a4ad94a538bb87e5016e7d5ce0bd434b
commit-date: 2022-11-19
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```
